### DISCUSSION TITLE
EditorContextService and AutoCompleteWidget - what are they good for?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#605] [Q&A] [answered]
I try to reimplement the TaskEditor UIExtension from the Workflow Example to get a more compact version to direct edit the label and some other details of an element within the graphical view. To avoid that I reinvent the wheel, can you explain a little bit what the classes `EditorContextService` and `AutoCompleteWidget` are good for?
The `EditorContextService`  looks promising like something handling the situation if I double click on a element with a label to switch into the edit mode. The AutoCompleteWidget seems to provide the input html element. But it also seems - [as you mentioned before](https://github.com/eclipse-glsp/glsp/discussions/591#discussioncomment-2374380) - it is an oversized version of a widget just to edit a label?

I guess in the glsp-client package there are a lot of interesting classes which can help me a lot if I could understand what they are good for. Maybe you find some time and just add one short comment into the head of each class of this package?

**Comment by planger** [comment, marked as answer]
The `EditorContextService` is a central injectable component that gives read-only access to certain aspects of the diagram, such as the currently selected elements, the model root (GModel), the edit mode, the latest position of the mouse in the diagram. It has been introduced for two main reasons:
1. to simplify accessing the model root and the current selection from components that are not commands, because in Sprotty only those had convenient access to that data
2. to conveniently create an `EditorContext`, which is a context object sent as part of several actions to the server to describe the current state of the editor (selection, last mouse position, etc.).

The `AutoCompleteWidget` is a reusable UI element that provides a text input supporting auto-completion, validation, validation messages, etc. In many adopter projects, this is used for inline editing or more complex label editing. An example is available in the workflow diagram: https://www.eclipse.org/glsp/gallery/#customcontextactions
![customcontextactions](https://user-images.githubusercontent.com/588090/160906289-a7229add-269e-4fb0-a0c2-95b6d3559a8b.gif)

I added these descriptions as documentation into the source code: https://github.com/eclipse-glsp/glsp-client/pull/173

For plain label editing (which is not as feature-rich as the auto-complete widget, e.g. no auto-completion), you'd just have to enable the label edit feature for a particular label type (see [workflow example](https://github.com/eclipse-glsp/glsp-client/blob/50434080a9890d7e9087fb5b561d5cea70f0230f/examples/workflow-glsp/src/di.config.ts#L58)):
```typescript
configureModelElement(context, 'label:heading', SLabel, SLabelView, { enable: [editLabelFeature] });
```

And handle the label editing actions on the server in the diagram module (see [workflow example](https://github.com/eclipse-glsp/glsp-server/tree/master/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow)):
```java
   @Override
   protected void configureOperationHandlers(final MultiBinding<OperationHandler> binding) {
      super.configureOperationHandlers(binding);
      binding.add(ApplyLabelEditOperationHandler.class);
   }

   @Override
   protected Class<? extends LabelEditValidator> bindLabelEditValidator() {
      return WorkflowLabelEditValidator.class;
   }
```

**Reply by rsoika** [reply]
Thanks a lot. That's making sense now. I always thought the edit label feature is based on AutoCompleteWidget. This was confusing me. But now it's clear.
